### ISSUE TITLE
Add JSON as a Custom Property data type

### ISSUE THREAD

**Issue by jjppof** [mapeditor/tiled#3186]
Hello!

Would be nice to have JSON data type when creating a Custom Property. JSON today is so popular that I believe that this feature would be valuable.

Thanks!

**Comment by bjorn** [comment]
I'm just wondering, but what would this JSON type change in practice, compared to using a plain string property, which also allows you to open a multi-line text editor dialog to edit the value?

Is this is about storing the value not as a string, but as actual JSON as part of a map or tileset saved in the JSON format?

**Comment by jjppof** [comment]
> Is this is about storing the value not as a string, but as actual JSON as part of a map or tileset saved in the JSON format?

Yeah, that's one reason.  The JSON data could come formatted in the map file which helps with quick edits. Also, the JSON could be validated on the property value set.

Another reason would be to distinguish what is a string from JSON, if I have two custom properties whether only their names are not enough to distinguish the type, I can't know which type is it without adding further information.

Also, but not necessary, JSON text highlight in the text editor would be the icing on the cake.

**Comment by eishiya** [comment]
With custom property types being an upcoming feature, I don't think a dedicated data type that's effectively a subtype of string should be necessary. Instead, I think something like a propertyEdited signal would be more robust, allowing people to hook up custom validation scripts. JSON validation in particular should be easy to do as a script, since JSON.parse(string) is available and returns fairly useful error messages when it encounters syntax errors.

Some syntax highlighting options in the string editor would also be great, and perhaps the custom property editor could have a field to allowing one to set default syntax highlighting preferences for each type, or it could just have some sort of basic "universal" highlighting that just makes it easy to match parentheses, brackets, and braces and highlights numbers, without any highlighting of language-specific keywords.
